### PR TITLE
Added bookmarks support

### PR DESCRIPTION
**Pull request by zBritva**
Added bookmarks support

### CODE REVIEW

File: src/sampleSlicer.ts [hunk 1]
```diff
@@ -293,6 +284,115 @@ module powerbi.extensibility.visual {
...
+
+            return false;
+        }
+
+        private restoreFilter(data: SampleSlicerData) {
+            let restoredFilter: IAdvancedFilter =
+            FilterManager.restoreFilter(data && data.slicerSettings.general.filter) as IAdvancedFilter;
+            if (restoredFilter) {
+                restoredFilter.target = this.getCallbacks().getAdvancedFilterColumnTarget();
+                if (this.checkFilter(restoredFilter) || true) {
```
Comment (ignatvilesov) [inline src/sampleSlicer.ts]: It is always true. Do we need this condition here?

File: src/sampleSlicer.ts [hunk 2]
```diff
@@ -293,6 +284,115 @@ module powerbi.extensibility.visual {
...
+                            });
+                        }
+                        if (operator === "GreaterThanOrEqual" || operator === "GreaterThan") {
+                            restoredFilter.conditions.push({
+                                operator: "LessThan",
+                                value: value.max
+                            });
+                        }
+                    }
+                    console.log('Restored', restoredFilter.conditions);
```
Comment (ignatvilesov) [inline src/sampleSlicer.ts]: Please remove the console.log

File: src/sampleSlicer.ts [hunk 3]
```diff
@@ -783,16 +888,15 @@ module powerbi.extensibility.visual {
...
         /**
          *  Callbacks consumed by the SelectionBehavior class
          * */
         private getCallbacks(): SampleSlicerCallbacks {
             let callbacks: SampleSlicerCallbacks = {};
 
             callbacks.applyAdvancedFilter = (filter: IAdvancedFilter): void => {
-                this.visualHost.applyJsonFilter(filter, "general", "filter");
+                this.visualHost.applyJsonFilter(filter, "general", "filter", FilterAction.merge );
+                this.currentFilter = filter;
```
Comment (ignatvilesov) [inline src/sampleSlicer.ts]: Why do we keep the filter if it's already stored in metadata?

File: src/selectionBehavior.ts [hunk 4]
```diff
@@ -110,57 +103,21 @@ module powerbi.extensibility.visual {
...
         public clearAllDiscreteSelections() {
-
+            try {
             /* update state to clear all selections */
-            this.selectionHandler.handleClearSelection();
+                if (this.selectionHandler) {
+                    this.selectionHandler.handleClearSelection();
+                }
+            }
+            catch (ex) {
```
Comment (ignatvilesov) [inline src/selectionBehavior.ts]: Please remove try - catch or log the error properly by console.error or throw ex

File: src/selectionBehavior.ts [hunk 5]
```diff
@@ -199,6 +156,7 @@ module powerbi.extensibility.visual {
 
             let filter: IAdvancedFilter = new window['powerbi-models'].AdvancedFilter(target, "And", conditions);
             this.callbacks.applyAdvancedFilter(filter);
+            console.log('saved', filter.conditions);
```
Comment (ignatvilesov) [inline src/selectionBehavior.ts]: would be better to remove

Comment (ignatvilesov) [reply src/selectionBehavior.ts]: Please also add the Travis config file to fix Travis CI build

File: package.json [hunk 6]
```diff
@@ -35,7 +36,7 @@
     "@types/nouislider": "9.0.0",
     "jasmine": "2.6.0",
     "jasmine-jquery": "2.1.1",
-    "powerbi-visuals-tools": "1.7.0",
+    "powerbi-visuals-tools": "1.11.2",
```
Comment (Guy-Moses) [inline package.json]: Isn't the highest version v1.11.1?

Comment (zBritva) [reply package.json]: @Guy-Moses no, we fixed `--install-cert` command in `1.11.2` There is details: https://github.com/Microsoft/PowerBI-visuals-tools/issues/166

File: doc/AddingBookmarksSuppoprt.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
+# Adding bookmarks support to the project
+
+Main Documentation about bookmarks can be found [here](https://github.com/Microsoft/PowerBI-visuals/blob/master/Tutorial/BookmarksSupport.md).
```
Comment (AviSander) [inline doc/AddingBookmarksSuppoprt.md]: The main documentation

File: doc/AddingBookmarksSuppoprt.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+To add bookmarks support for the visual, you should update `powerbi-visuals-utils-interactivityutils` to version 3.0.0 or higer.
```
Comment (AviSander) [inline doc/AddingBookmarksSuppoprt.md]: you should update to version 3.0.0 or higher of `powerbi-visuals-utils-interactivityutils`.

File: doc/AddingBookmarksSuppoprt.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+In update `updateOnRangeSelectonChange` method the visual create AdvancedFilter and apply filter with one or two conditions.
```
Comment (AviSander) [inline doc/AddingBookmarksSuppoprt.md]: +In the `updateOnRangeSelectonChange` method the visual creates an AdvancedFilter and applies the filter with one or two conditions.

File: doc/AddingBookmarksSuppoprt.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+And calls `applyAdvancedFilter`, where `applyAdvancedFilter` [pass](https://github.com/Microsoft/powerbi-visuals-sampleslicer/blob/master/src/sampleSlicer.ts#L795) call to `this.visualHost.applyJsonFilter`.
```
Comment (AviSander) [inline doc/AddingBookmarksSuppoprt.md]: calls `this.visualHost.applyJsonFilter`.

File: doc/AddingBookmarksSuppoprt.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+In each update the visual [check and restores](https://github.com/Microsoft/powerbi-visuals-sampleslicer/pull/6/files#diff-5929da3be6a696fb9df5e3571baceb52R356) the persisted filter by using [`FilterManager.restoreFilter`](https://github.com/Microsoft/PowerBI-visuals/blob/master/Tutorial/BookmarksSupport.md#visuals-with-filter).
```
Comment (AviSander) [inline doc/AddingBookmarksSuppoprt.md]: [inspects and restores]

File: doc/AddingBookmarksSuppoprt.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+```typescript
+private restoreFilter(data: SampleSlicerData) {
+    // restore advanced filter from vsual properties
```
Comment (AviSander) [inline doc/AddingBookmarksSuppoprt.md]: visual

File: doc/AddingBookmarksSuppoprt.md [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    let restoredFilter: IAdvancedFilter =
+    FilterManager.restoreFilter(data && data.slicerSettings.general.filter) as IAdvancedFilter;
+    // if filter was persisted, the visual get conditions of advanced filter
+    if (restoredFilter) {
+        restoredFilter.target = this.getCallbacks().getAdvancedFilterColumnTarget();
+        // reset to default
+        // change value to correspond the filter values
```
Comment (AviSander) [inline doc/AddingBookmarksSuppoprt.md]: modify the values to match the filter values

File: doc/AddingBookmarksSuppoprt.md [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+        // in some case we can get value with one condition only
```
Comment (AviSander) [inline doc/AddingBookmarksSuppoprt.md]: // in some cases we can receive values with only one condition

File: doc/AddingBookmarksSuppoprt.md [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+        if (restoredFilter.conditions.length === 1) {
+            let value: {
+                max?: any,
+                min?: any
+            } = {};
+
+            // get max and min values in dataset
```
Comment (AviSander) [inline doc/AddingBookmarksSuppoprt.md]: // get min and max values in the dataset

File: doc/AddingBookmarksSuppoprt.md [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+            let convertedValues = data.slicerDataPoints.map( (dataPoint: SampleSlicerDataPoint) => +dataPoint.category );
+            value.min = d3.min(convertedValues);
+            value.max = d3.max(convertedValues);
+
+            // if some conditions is absent, the visuals adds the condition with correspond value
```
Comment (AviSander) [inline doc/AddingBookmarksSuppoprt.md]: // if some conditions is missing, the visual adds the condition with matching value

File: doc/AddingBookmarksSuppoprt.md [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+            }
+            if (operator === "GreaterThanOrEqual" || operator === "GreaterThan") {
+                restoredFilter.conditions.push({
+                    operator: "LessThan",
+                    value: value.max
+                });
+            }
+        }
+
+        // create ValueRange object to apply current filter state to slicer visual
```
Comment (AviSander) [inline doc/AddingBookmarksSuppoprt.md]: // create ValueRange object to apply current filter state to the slicer visual

File: doc/AddingBookmarksSuppoprt.md [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+        // change visual state of slicer
+        this.behavior.scalableRange.setValue(rangeValue);
+        // change visual state of text boxes
+        this.onRangeInputTextboxChange(rangeValue.min.toString(), RangeValueType.Start);
+        this.onRangeInputTextboxChange(rangeValue.max.toString(), RangeValueType.End);
+    }
+}
+```
+Resuming: in this method the visual restores `restoredFilter` object, parse conditions for restoring saved values of slicer and apply values to slicer and text boxes.
```
Comment (AviSander) [inline doc/AddingBookmarksSuppoprt.md]: Resuming: in this method the visual restores the `restoredFilter` object, parses conditions for restoring saved values of the slicer and applies values to slicer and text boxes.

File: doc/AddingBookmarksSuppoprt.md [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+
+You don't need persist the filter in the visual properties. And `persistSelectionState`, `restorePersistedRangeSelectionState` and other methods were removed.
```
Comment (AviSander) [inline doc/AddingBookmarksSuppoprt.md]: You should not persist...

File: doc/UsingInteractivityUtils.md [hunk 20]
```diff
@@ -8,17 +8,14 @@ For <b>discrete</b> cross-visual data-point selection the Sample Slicer visual r
 
 ISelectionHandler holds the state of all discrete (possibly multiple) data-point selections. The handler is updated on each data-point selection event (each mouse click) and does NOT automatically propagate the event to the hosting application. The selection state is only propagated to the host when the method ISelectionHandler.applySelectionFilter() is invoked. 
 
-Below is the code executed on each slicer mouse click. The handler is updated with the select/unselect data point event and the complete discrete selection state is flushed to the hosting application. Additionally, the selection state is persisted to the visual's properties so the next time the visual is loaded the selection can be restored. 
+Below is the code executed on each slicer mouse click. The handler is updated with the select/unselect data point event and the complete discrete selection state is flushed to the hosting application. Additionally, the selection state is persisted to the visual's properties as applyed filter, so the next time the visual is loaded the selection can be restored.
```
Comment (AviSander) [inline doc/UsingInteractivityUtils.md]: as the applied filter,

File: src/selectionBehavior.ts (hunk 5 above)
Comment (vizakhar) [inline src/selectionBehavior.ts]: I see some sections of code related to the filter/selection persistance are being removed. Will the selection be maintained if you close and later on re-open a report?

**Pull request merged**